common/thorhelper/thorfile.cpp

```diff
@@ -100,3 +101,257 @@ IHThorDiskReadArg * createWorkUnitReadArg(const char * filename, IHThorWorkunitR
+    unsigned nodeSize = attrs.getPropInt64("@nodeSize", defaultNodeSize);
+    if (attrs.hasProp("@numLeafNodes"))
+    {
+        result.knownLeafCount = true;
+        result.numLeafNodes = attrs.getPropInt64("@numLeafNodes");
+        result.numBlobNodes = attrs.getPropInt64("@numBlobNodes");
+        result.numBranchNodes = attrs.getPropInt64("@numBranchNodes");
+        result.sizeDiskLeafs = result.numLeafNodes * nodeSize;
+        result.sizeDiskBlobs = result.numBlobNodes * nodeSize;
+        result.sizeDiskBranches = result.numBranchNodes * nodeSize;
+    }
+    else
+    {
+        Owned<IDistributedFilePartIterator> parts = file->getIterator();
+        ForEach(*parts)
+        {
+            IDistributedFilePart & curPart = parts->query();
+            //Don't include the TLK in the extended information
+            if (isPartTLK(&curPart))
+                continue;
+
+            IPropertyTree & partAttrs = curPart.queryAttributes();
+            offset_t branchOffset = partAttrs.getPropInt64("@offsetBranches");
+            offset_t compressedSize = partAttrs.getPropInt64("@compressedSize");
+            offset_t uncompressedSize = partAttrs.getPropInt64("@size");
+            //New builds publish compressed and uncompressed sizes, old builds only publish compressed as @size
+            if (compressedSize == 0)
+                compressedSize = uncompressedSize;
+
+            if (branchOffset != 0)
+            {
+                offset_t sizeLeafs = branchOffset - nodeSize;                        // Assume all leaf nodes except leading header (no blobs)
+                offset_t sizeBranches = (compressedSize - branchOffset) - nodeSize;  // An over-estimate - trailing header, meta, blooms
+                result.numLeafNodes += sizeLeafs / nodeSize;
+                result.numBranchNodes += sizeBranches / nodeSize;
+                result.sizeDiskLeafs += sizeLeafs;
+                result.sizeDiskBranches += sizeBranches;
+            }
+            else
+            {
+                //A single leaf node...
+                result.numLeafNodes += 1;
+                result.sizeDiskLeafs += nodeSize;
+            }
+        }
+    }
+
+    unsigned keyLen = attrs.getPropInt64("@keySize");
+    result.sizeOriginalBranches = result.numLeafNodes * (keyLen + 8);
+    offset_t compressedSize = attrs.getPropInt64("@compressedSize");
+    offset_t uncompressedSize = attrs.getPropInt64("@size");
+    //size only corresponds to the uncompressed size if both size and compressedSize are filled in.
+    if ((uncompressedSize != 0) && (compressedSize != 0))
+        result.sizeOriginalData = uncompressedSize;
+
+    //The following will depend on the compression format - e.g. if compressed searching is implemented
+    result.sizeMemoryBranches = result.sizeOriginalBranches;
+
+    //Could calculate from the size of the raw record (with blobids) multiplied by the count.
+    if ((result.numBlobNodes == 0) && result.sizeOriginalData)
```

Comment:
@jakesmith a question.  Currently the value filled in in @size is the size of the rows sent to the builder, but this doesn't include the size of any blobs.  That means this calculation is wrong.  There are two potential fixes:

* Always go through the 1st branch - since blobs aren't included they don't affect the in memory leaf size
* Add the size of any blobs to the @size reported for a file.

Making the second change means that the compression ratio will be more accurate.  We could also record @sizeOriginalBlobs in the meta, but is that overkill?

Reply:
I can't think of any good reason not to include the blob size in @size too, and to record sizeOriginalBlobs, which I think could be useful info. too - other than effort.